test(primitives): Extract commonalities from market and storage-provider pallet benchmarks, and use real proof data

Jinxit opened the pull request:
### Description

A problem arising from using real proofs for benchmarking the storage-provider pallet is that the runtime must use the real proof verifier, which means the market pallet benchmarks break. While the setup created for storage-provider _could_ be copied across, it feels a bit more maintainable to:
1. Extract commonalities and reuse them.
2. Create a reproducible (and deterministic) way to generate the data.

This PR attempts to move test setup code, mostly for benchmarks, which is currently duplicated between the market and storage-provider pallets. I didn't try to extract _everything_, I was mostly aiming to get it to work to not make an already large PR any larger. There's a new CLI command to `storage-provider-client` called `benchmark-data` to generate code, keys and proofs which are then embedded in the benchmark binary. I've also created traits for things which are common to the respective pallets' configs, such as `CurrencyProvider`, `BalanceOf`, and traits for each pallet's config itself - since that is kind of part of the (internal) interfaces of the pallets.

**This is the big one**, so I'm very much open to comments for improvements. Going commit by commit might not _compile_, but the changes are grouped on a best-effort basis to make reviewing them easier. `be6fb5d` is a bit of a mess though, I'll admit.

There is some overlap with what this is doing and the downloading of parameters (`just download-params`), but I haven't really made any attempt to unify the two approaches.

### Important points for reviewers

As a side effect of creating `MarketProvider` and `StorageProviderProvider` (what a name, it's like I'm back in Java-land), the `#[pallet::constant]` tags have disappeared from the constants which are used by the other pallet, respectively. This means the constants are no longer a part of the pallet metadata, which I _think_ is fine for now. The end goal should probably be to merge the two pallets anyway and move stuff back out of `primitives`.

### Checklist

- [x] Are there important points that reviewers should know?
  - [x] Less pallet metadata
- [x] Make sure that you described what this change does.
- [x] If there are follow-ups, have you created issues for them?
  - [x] #739 
- [x] Have you tested this solution?
- [x] Were there any alternative implementations considered?
  - [x] Copy-paste the setup
  - [x] Merge storage-provider and market (maybe later?) 
- [x] Did you document new (or modified) APIs?

Jinxit committed:
feat(storage-provider-client): Store multiple proofs as a simple array instead of scale encoding a vector

Signed-off-by: Lucas Åström <lucas@eiger.co>

Jinxit committed:
feat(storage-provider-client): Add a command for generating proof data for benchmarks

Signed-off-by: Lucas Åström <lucas@eiger.co>

Jinxit committed:
feat(primitives): Extract and implement traits for market and storage-provider configs

Signed-off-by: Lucas Åström <lucas@eiger.co>

Jinxit committed:
feat(primitives): Move deal structs from market to primitives

Signed-off-by: Lucas Åström <lucas@eiger.co>

Jinxit committed:
test(primitives): Extract common functions used in both market and storage-provider benchmarks, and use the same real proof data in both

Signed-off-by: Lucas Åström <lucas@eiger.co>

Jinxit committed:
test(primitives): Embed benchmark proof data

Signed-off-by: Lucas Åström <lucas@eiger.co>

th7nder commented on storage-provider/client/src/commands/proofs.rs:
```diff
@@ -527,6 +597,228 @@ fn post(
+    let sector_size = seal_proof.sector_size();
+    let porep_params_path = params_root.join(format!("{sector_size}.{POREP_PARAMS_EXT}"));
+    let porep_params_vk_path = params_root.join(format!("{sector_size}.{POREP_VK_EXT_SCALE}"));
+    if !tokio::fs::try_exists(&porep_params_vk_path).await? {
```
This should fail instead of generate, we should always use those shared params that are (**should**) be stored on blob.

Jinxit replied:
This was written before the downloading of params so it's generating a completely separate set - I'll switch it over.

Jinxit replied:
Actually, thinking about this again, shouldn't it generate if they are missing? If we're using this as a command to run and then commit to the repo (or to the blob store), then this should be how that data is generated in the first place. This command shouldn't be run on developer machines, really, outside of developing the command itself.

th7nder replied:
I'd agree about the generation, but to generate them properly you need to have a huge-ass machine and then update them in lots of places. Like the repo, chain, genesis, etc. 
I think this is not a trivial step and needs to be done **explicitly**.

th7nder commented on primitives/src/test_data.rs:
```diff
@@ -0,0 +1,1662 @@
+            // This code has been generated by `target/release/polka-storage-provider-client proofs benchmark-data --sector-size 2KiB examples/test-data-big.car`
+            SectorSize::_2KiB => BenchmarkData {
+                storage_provider_name: "//StorageProvider",
+                verifying_key: include_bytes!("../../target/bench/params/2KiB.porep.vk.scale"),
```
We need to store verifying key as well as generated proofs in the repo, somewhere committed. If we don't, CI will fail.
Params cannot be stored on the repo, as they're too big.

I'd suggest the workflow looks like this:
- we generate the bench data for 1GiB sectors (both proofs, vks and stuff) on the beefy machine, by running the command `benchmark-data`
- upload this as a backup to our blob storage (i'm doing it now)
- commit the generate benched data to the repo (as well as the benchmarks)

Jinxit replied:
Sounds good, I'll adjust the paths accordingly and add it to the download script.

th7nder commented on storage-provider/client/src/commands/proofs.rs:
```diff
@@ -527,6 +597,228 @@ fn post(
+        .join(sector_size.to_string());
+    tokio::fs::create_dir_all(&proofs_root).await?;
+
+    for sector_number in 0..MAX_SECTORS_PER_CALL {
```
This `MAX_SECTORS_PER_CALL` is probably to be adjusted. I think that with 1GiB sectors we'll be able to `ProveCommit` only 1 per call, PreCommit is much different, it's not that compute heavy, but still.

To generate 20 seperate proofs for 1GiB, it'll take like 5 hours (1 proof -> ~20minutes). 
Do we wanna do that if we'll accept one proof for prove commit anyways?

Jinxit replied:
If we're planning on lowering the max sectors (even to 1), then yeah I agree this is a good time to do it.

th7nder commented on pallets/market/src/lib.rs:
```diff
@@ -58,49 +57,21 @@ pub mod pallet {
-
     #[pallet::config]
-    pub trait Config: frame_system::Config {
+    pub trait Config: frame_system::Config + CurrencyProvider + MarketProvider {
```
This seems fishy, but hey, as long as it works.

Jinxit replied:
It's kinda the same thing that we're already doing by inheriting `frame_system::Config` to get access to `AccountId` and `BlockNumber` etc.

th7nder replied:
Ok

th7nder commented on pallets/market/benchmarks/src/mock.rs:
```diff
@@ -78,44 +81,53 @@ parameter_types! {
-    pub const WPoStPeriodDeadlines: u64 = 3;
-    pub const WPostChallengeWindow: BlockNumber = 2 * MINUTES;
-    pub const WPostChallengeLookBack: BlockNumber = MINUTES;
+    pub const WPoStPeriodDeadlines: u64 = 10;
+    pub const WPoStProvingPeriod: BlockNumber = 40 * MINUTES;
+    pub const WPoStChallengeWindow: BlockNumber = 4 * MINUTES;
+    pub const WPoStChallengeLookBack: BlockNumber = MINUTES;
     pub const MinSectorExpiration: BlockNumber = 5 * MINUTES;
```
Why did we change those values anyways?
'Mainnet' ones (the place where we actually need benchmarking) are a bit [different](https://github.com/eigerco/polka-storage/blob/817826c86d95aeea7ee6121c831d910b015ee9f9/runtime/src/configs/mod.rs#L316),.

Btw. the current `testnet` is more like a `devnet`.

Jinxit replied:
I don't remember the exact error but I think I ran into timing issues with the previous values, so I matched what we have in the storage-provider pallet.

th7nder replied:
I don't mind, i don't think those values are affecting benchmarks in any meaningful way.

th7nder commented on primitives/src/test_data.rs:
```diff
@@ -0,0 +1,1662 @@
+extern crate alloc;
```
shouldn't test_data be behind some feature flag?
It'll pull a lot of stuff into the WASM runtime when when we're doing 'include_bytes'

Jinxit replied:
It's behind a `#[cfg(feature = "testing")]` in the `lib.rs` file.

th7nder replied:
Cool

th7nder commented on primitives/src/test_data.rs:
```diff
@@ -0,0 +1,1662 @@
+
+    pub fn load(sector_size: SectorSize) -> BenchmarkData<T> {
+        match sector_size {
+            // DO NOT MODIFY
```
Btw. shouldn't we require only one sector_size here to be loaded?
Like if we need benchmarks only for 1GiB/8MiB to not pollute the binary?

Jinxit replied:
You mean with features or something? I mostly started like this when going back and forth between generating data and running it, it was convenient to easily swap them out.

th7nder replied:
I guess it was bothering me that we'll only use SectorSize 1GiB for benchmarks and have multiple sector sizes in this switch which are not necessary.

Jinxit replied:
Yeah it's a good point, I think I'll switch it out (ha) to just returning the 1GB data as soon as I get that working.

th7nder reviewed (changes requested):
Good stuff! In this first pass I just focused on the high-level functionality and concepts, not at the trait choices/code design.

Jinxit committed:
fix(primitives): Reduce the number of sectors per call to 1

Signed-off-by: Lucas Åström <lucas@eiger.co>

cernicc commented on storage-provider/client/src/commands/proofs.rs:
```diff
@@ -237,15 +317,16 @@ async fn calculate_piece_commitment(input_path: PathBuf) -> Result<(), CliError>
+    output_path: Option<impl AsRef<Path>>,
     seal_proof: RegisteredSealProof,
 ) -> Result<(), CliError> {
     let output_path = if let Some(output_path) = output_path {
```
I think it is better to accept an owned type if the ownership is needed by the function  logic. The caller can then decide if another allocation is needed or if the ownership can just be transferred to the function.

In this example the owned value is immediately created from the reference. This is a good indication of the point above.

Jinxit replied:
It was less about not cloning and more about developer experience in an earlier commit where I was using a `TempDir` directly which implements `AsRef<Path>`. That changed along the way, and now it's just a `PathBuf` on both call sites.

I'm up for changing it, but I don't think performance is the right reason. The command takes minutes to run after all. But writing simpler code I can agree to.

Jinxit replied:
I think I prefer the `AsRef<Path>` version, but here's the less-clone version: 7acdd0e33abe11f4421adf4ae2a5ba4419314591

cernicc commented on storage-provider/client/src/commands/proofs.rs:
```diff
@@ -130,6 +144,59 @@ pub enum ProofsCommand {
         /// CID - CommR of a replica (output of `porep` command)
         comm_r: String,
     },
+    /// Generates PoRep params, PoSt params, proofs and commitments used for benchmarking.
```
I have a feeling like we should hide everything required for benchmarks behind some feature flag. What do you think?

Jinxit replied:
I agree. At first I considered `runtime-benchmarks` but I'm not sure whether that's already too overloaded with the _actual_ benchmarking code.

Jinxit replied:
@jmg-duarte @th7nder
What do you think I should call it?

th7nder replied:
I don't mind if this is not behind a feature flag, but `benchmarking` seems reasonable (?)

cernicc commented on primitives/src/configs.rs:
```diff
@@ -0,0 +1,46 @@
+pub type BalanceOf<T> = <<T as CurrencyProvider>::Currency as Currency<
+    <T as frame_system::Config>::AccountId,
+>>::Balance;
+pub trait CurrencyProvider: frame_system::Config {
```
Missing new line before the trait

cernicc reviewed (commented):
This benchmarking looks like a rocket science :D Left some comments. Good job figuring this out.

Jinxit committed:
fix(market,storage-provider): Remove linear scaling on number of sectors in benchmarks

Polkadot doesn't like `Linear<1, 1>`, it triggers an out-of-bounds panic.

Signed-off-by: Lucas Åström <lucas@eiger.co>

Jinxit committed:
refactor(proofs): Shuffle around paths for proof params and keys, and commit example proofs for 8MiB and 1GiB sectors

Signed-off-by: Lucas Åström <lucas@eiger.co>

Jinxit committed:
fix(market,storage-provider): Use the correct lengths for some bounded vecs

Signed-off-by: Lucas Åström <lucas@eiger.co>

Jinxit commented:
I didn't address all comments yet, but I've now updated to using the 1GB proof data and only doing 1 sector at a time. For now I've left it as a `BoundedVec` with a length of 1 to not change APIs everywhere, there's already enough in this PR.

Please give some extra attention to b587a4ed0c9545723ac986f8d7dfbfc7e82b2252 - I'm not 100% sure I replaced the right constants, but I think some of them were wrong in the first place (and happened to fit).

Jinxit committed:
fix(primitives): Don't automatically generate proof params if they are missing

Signed-off-by: Lucas Åström <lucas@eiger.co>

Jinxit committed:
fix: Formatting

Signed-off-by: Lucas Åström <lucas@eiger.co>

Jinxit committed:
refactor(sp-client): Use PathBuf instead of Path when function clones it anyway

Signed-off-by: Lucas Åström <lucas@eiger.co>

jmg-duarte commented on Cargo.toml:
```diff
@@ -97,6 +97,7 @@ pairing = "0.23"
 polkavm = "0.9.3"
 polkavm-derive = "0.9.1"
 polkavm-linker = "0.9.2"
+proc-macro2 = { version = "1.0.93" }
```
This looks kinda sus, version unification?

Jinxit replied:
I added it as a dependency for storage-provider-client, where I use a `proc_macro2::TokenStream` to generate the output code.

jmg-duarte commented on pallets/market/benchmarks/src/benchmarking.rs:
```diff
@@ -231,159 +162,124 @@ mod benchmarks {
+            .map(|p| p.proposal.total_storage_fee().unwrap())
+            .sum::<u128>()
+            .try_into()
+            .unwrap();
```
I get why you're doing the u128 -> u32 -> Balance, but maybe you can just do u128 -> Balance through the UniqueSaturatedFrom

Because:
```
Balance: AtLeast32BitUnsigned
AtLeast32BitUnsigned: AtLeast32Bit
AtLeast32Bit: BaseArithmetic
BaseArithmetic: UniqueSaturatedFrom<u128>
```

https://paritytech.github.io/polkadot-sdk/master/sp_arithmetic/traits/trait.BaseArithmetic.html

jmg-duarte commented on pallets/market/benchmarks/src/benchmarking.rs:
```diff
@@ -231,159 +162,124 @@ mod benchmarks {
+        run_to_block::<T>(
+            expiration_block
+                .try_into()
+                .unwrap_or_else(|_| panic!("failed to convert expiration block to block number")),
```
What is the difference of this and `expect`?

Jinxit replied:
`expect` requires a `Debug` implementation.

jmg-duarte commented on pallets/market/benchmarks/src/mock.rs:
```diff
@@ -78,44 +81,53 @@ parameter_types! {
-    pub const MinDealDuration: u64 = 5 * MINUTES;
-    pub const MaxDealDuration: u64 = 180 * MINUTES;
-    pub const PreCommitChallengeDelay: BlockNumber = 1 * MINUTES;
+    // NOTE: Changed from Testnet's 0 to 10 to match the storage-provider client's `porep` command.
+    pub const PreCommitChallengeDelay: BlockNumber = 10;
+    pub const MinDealDuration: u64 = 2 * MINUTES;
+    pub const MaxDealDuration: u64 = 30 * MINUTES;
```
We should be testing with the values that are going to make it to production. I'm not sure what those are but this comment doesn't inspire much confidence either.

We might just need to make it clear what "testnet" means for our project, etc

cc @th7nder

th7nder replied:
I'd say we can leave it be in this PR, as matching those values to the mainnet ones will require changes across many things. Just create a follow-up I'd say.

Jinxit replied:
#824

jmg-duarte commented on pallets/market/benchmarks/src/mock.rs:
```diff
@@ -78,44 +81,53 @@ parameter_types! {
+    pub const MaxSectorExpiration: BlockNumber = 360 * MINUTES;
     pub const SectorMaximumLifetime: BlockNumber = 120 * MINUTES;
-    pub const MaxProveCommitDuration: BlockNumber = 5 * MINUTES;
+    // NOTE: Changed from Testnet's 5 to 15 since the pre-commit delay is 10 blocks.
```
Same as above

th7nder replied:
#824

jmg-duarte commented on primitives/src/test_data.rs:
```diff
@@ -0,0 +1,246 @@
+    pub comm_p: Commitment<CommP>,
+    pub sectors: Vec<SectorData>,
+    _phantom: PhantomData<T>,
+}
```
Move this to be close to the impl

jmg-duarte commented on pallets/storage-provider/benchmarks/src/benchmarking.rs:
```diff
@@ -0,0 +1,299 @@
+    BlockNumberFor<T>: From<u64> + Into<u64>,
+{
+    let state = SpPallet::<T>::storage_providers(sp_id).unwrap();
+    let balance: BalanceOf<T> = (n * 1_u32).into();
```
What is the `* 1_u32` doing?

Jinxit replied:
I think it used to be something else, so it's just a leftover that I forgot to remove.

jmg-duarte reviewed (commented):
This PR is too big for me to provide a meaningful review. I'd prefer having it split into multiple parts but I don't want to discard the reviews that were already done.

Next time, I'm rejecting.

Jinxit committed:
test(market): Convert straight to Balance without intermediate u32 step

Signed-off-by: Lucas Åström <lucas@eiger.co>

th7nder commented on pallets/storage-provider/benchmarks/src/benchmarking.rs:
```diff
@@ -0,0 +1,299 @@
+        data.verifying_key.to_vec(),
+    ));
+
+    run_to_block::<T>(15.into());
```
those 'magic values' will bite us when we change to the 'mainnet params'

Jinxit replied:
I didn't replace it in all tests etc, but at least in the benchmarks those numbers are now based on the pallet constants.

th7nder replied:
good, not saying you should do this now - just a note - resolving

th7nder commented on primitives/src/test_data.rs:
```diff
@@ -0,0 +1,246 @@
+        // This code has been generated by `./target/release/polka-storage-provider-client proofs benchmark-data --sector-size 1GiB examples/big_file_184k.car`
+        BenchmarkData {
+            storage_provider_name: "//StorageProvider",
+            verifying_key: include_bytes!("../../examples/benchmark/keys/1GiB.porep.vk.scale"),
```
nit: I'd say overall the `examples/` directory stopped being useful for purposes of both bechmarking and testing.
1. The keys we're using in benchmarks are also used in genesis config, so it's a bit odd that genesis config uses something from `examples/benchmark/keys`.

`test-fixtures` or something similar would be more fitting, but this is just naming. It's not like super important, it just creates confusion.

Jinxit replied:
Yeah, I agree. I kinda got lost a bit when I realized I don't currently have replacements for the _other_ proofs in `examples/` (specifically PoSt). Before then it was named `test-data`, and I wanted to get this PR done so I went with `benchmark` to distinguish the two.

th7nder commented on storage-provider/client/src/commands/proofs.rs:
```diff
@@ -527,6 +596,224 @@ fn post(
+    let sector_size = seal_proof.sector_size();
+    let porep_params_path = params_root.join(format!("{sector_size}.{POREP_PARAMS_EXT}"));
+    let porep_params_vk_path = keys_root.join(format!("{sector_size}.{POREP_VK_EXT_SCALE}"));
+    if !tokio::fs::try_exists(&porep_params_vk_path).await? {
```
```suggestion
    if !tokio::fs::try_exists(&porep_params_path) || !tokio::fs::try_exists(&porep_params_vk_path).await? {
```

th7nder reviewed (dismissed):
That's a piece of work which seems complete, great work! I wouldn't say any of those comments is blocking.

Jinxit committed:
refactor(primitives): Move BenchmarkData struct def to impl

Signed-off-by: Lucas Åström <lucas@eiger.co>

Jinxit committed:
test(market,sp): Replace magic values

Signed-off-by: Lucas Åström <lucas@eiger.co>

Jinxit committed:
fix(sp-client): Check for both params and verifying key

Signed-off-by: Lucas Åström <lucas@eiger.co>

Jinxit committed:
refactor: Move examples/benchmark to test-fixtures

Signed-off-by: Lucas Åström <lucas@eiger.co>
